// ngx-bootstrap-navbar/esm2020/lib/navbar.module.mjs
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { NgxNavbarCollapseComponent } from './navbar-collapse.component';
import { NgxNavbarDynamicExpandDirective } from './navbar-dynamic-collapse.directive';
import * as i0 from "@angular/core";
export class NgxNavbarModule {
}
NgxNavbarModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "13.0.0", ngImport: i0, type: NgxNavbarModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
NgxNavbarModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "12.0.0", version: "13.0.0", ngImport: i0, type: NgxNavbarModule, declarations: [NgxNavbarCollapseComponent, NgxNavbarDynamicExpandDirective], imports: [CommonModule], exports: [NgxNavbarCollapseComponent, NgxNavbarDynamicExpandDirective] });
NgxNavbarModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "13.0.0", ngImport: i0, type: NgxNavbarModule, imports: [[CommonModule]] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "13.0.0", ngImport: i0, type: NgxNavbarModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [NgxNavbarCollapseComponent, NgxNavbarDynamicExpandDirective],
                    imports: [CommonModule],
                    exports: [NgxNavbarCollapseComponent, NgxNavbarDynamicExpandDirective]
                }]
        }] });
